### PR TITLE
GH-2009: Updated the dependencies to support Node.js 10+.

### LINKED ISSUE #2009: Theia does not run with Node.js `v10.3.0`
Help wanted. See: https://github.com/theia-ide/theia/pull/3358#issuecomment-434978417
```
$ nvm use 10 && node --version && npm --version && yarn --version && python --version && git clone https://github.com/theia-ide/theia.git && cd theia && yarn && cd examples/browser && yarn run start
Now using node v10.3.0 (npm v6.1.0)
v10.3.0
6.1.0
1.5.1
Python 2.7.10
Cloning into 'theia'...
remote: Counting objects: 25608, done.
remote: Compressing objects: 100% (94/94), done.
remote: Total 25608 (delta 41), reused 42 (delta 18), pack-reused 25495
Receiving objects: 100% (25608/25608), 118.12 MiB | 2.32 MiB/s, done.
Resolving deltas: 100% (17318/17318), done.
yarn install v1.5.1
(node:5029) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
[3/5] 🚚  Fetching packages...
warning vscode-base-languageclient@0.0.1-alpha.5: The engine "vscode" appears to be invalid.
[4/5] 🔗  Linking dependencies...
warning " > istanbul-instrumenter-loader@3.0.1" has unmet peer dependency "webpack@^2.0.0 || ^3.0.0 || ^4.0.0".
warning "workspace-aggregator-b404bb6e-1773-4313-be3f-6c77f10d80b0 > @theia/application-manager > font-awesome-webpack@0.0.5-beta.2" has unmet peer dependency "font-awesome@>=4.3.0".
[5/5] 📃  Building fresh packages...
[1/11] ⠁ fibers:                               ^
[2/11] ⠁ phantomjs-prebuilt: Receiving...
[3/11] ⠁ dtrace-provider: ACTION binding_gyp_ndtp_target_build_ndtp .
[7/11] ⠁ node-pty: CXX(target) Release/obj.target/pty/src/unix/pty.o
error An unexpected error occurred: "/Users/akos.kitta/Desktop/theia/node_modules/fibers: Command failed.
Exit code: 127
Command: sh
Arguments: -c node build.js || nodejs build.js
Directory: /Users/akos.kitta/Desktop/theia/node_modules/fibers
Output:
gyp info it worked if it ends with ok
gyp info using node-gyp@3.6.2
gyp info using node@10.3.0 | darwin | x64
gyp http GET https://nodejs.org/download/release/v10.3.0/node-v10.3.0-headers.tar.gz
gyp http 200 https://nodejs.org/download/release/v10.3.0/node-v10.3.0-headers.tar.gz
gyp http GET https://nodejs.org/download/release/v10.3.0/SHASUMS256.txt
gyp http 200 https://nodejs.org/download/release/v10.3.0/SHASUMS256.txt
gyp info spawn /usr/bin/python
gyp info spawn args [ '/Users/akos.kitta/Desktop/theia/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/akos.kitta/Desktop/theia/node_modules/fibers/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/akos.kitta/Desktop/theia/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/Users/akos.kitta/.node-gyp/10.3.0/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/Users/akos.kitta/.node-gyp/10.3.0',
gyp info spawn args   '-Dnode_gyp_dir=/Users/akos.kitta/Desktop/theia/node_modules/node-gyp',
gyp info spawn args   '-Dnode_lib_file=/Users/akos.kitta/.node-gyp/10.3.0/<(target_arch)/node.lib',
gyp info spawn args   '-Dmodule_root_dir=/Users/akos.kitta/Desktop/theia/node_modules/fibers',
gyp info spawn args   '-Dnode_engine=v8',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.' ]
xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance

xcode-select: error: tool 'xcodebuild' requires Xcode, but active developer directory '/Library/Developer/CommandLineTools' is a command line tools instance

gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
  CXX(target) Release/obj.target/fibers/src/fibers.o
../src/fibers.cc:122:64: error: too few arguments to function call, expected at least 3, have 2
                return String::NewFromOneByte(isolate, (const uint8_t*)string);
                       ~~~~~~~~~~~~~~~~~~~~~~                                ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:2736:3: note: 'NewFromOneByte' declared here
  static V8_WARN_UNUSED_RESULT MaybeLocal<String> NewFromOneByte(
  ^
../src/fibers.cc:126:64: error: too few arguments to function call, expected at least 3, have 2
                return String::NewFromOneByte(isolate, (const uint8_t*)string);
                       ~~~~~~~~~~~~~~~~~~~~~~                                ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:2736:3: note: 'NewFromOneByte' declared here
  static V8_WARN_UNUSED_RESULT MaybeLocal<String> NewFromOneByte(
  ^
../src/fibers.cc:406:24: warning: 'Utf8Value' is deprecated [-Wdeprecated-declarations]
                                        String::Utf8Value stack(uni::Deref(that.isolate, fatal_stack));
                                                          ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:2821:5: note: 'Utf8Value' has been explicitly marked deprecated here
    V8_DEPRECATED(\"Use Isolate version\",
    ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8config.h:321:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated))
                            ^
../src/fibers.cc:433:80: error: no matching member function for call to 'NewInstance'
                                return uni::Return(uni::Deref(Isolate::GetCurrent(), tmpl)->GetFunction()->NewInstance(1, argv), args);
                                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:3862:44: note: candidate function not viable: requires single argument 'context', but 2 arguments were provided
  V8_WARN_UNUSED_RESULT MaybeLocal<Object> NewInstance(
                                           ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:3859:44: note: candidate function not viable: requires 3 arguments, but 2 were provided
  V8_WARN_UNUSED_RESULT MaybeLocal<Object> NewInstance(
                                           ^
../src/fibers.cc:621:14: error: no matching constructor for initialization of 'v8::TryCatch'
                                TryCatch try_catch;
                                         ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:8282:12: note: candidate constructor not viable: requires single argument 'isolate', but no arguments were provided
  explicit TryCatch(Isolate* isolate);
           ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:8402:3: note: candidate constructor not viable: requires 1 argument, but 0 were provided
  TryCatch(const TryCatch&) = delete;
  ^
../src/fibers.cc:628:13: warning: 'Compile' is deprecated [-Wdeprecated-declarations]
                                Script::Compile(uni::NewLatin1String(that.isolate, \"void 0;\"));
                                        ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:1234:10: note: 'Compile' has been explicitly marked deprecated here
  static V8_DEPRECATED(\"Use maybe version\",
         ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8config.h:321:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated))
                            ^
../src/fibers.cc:741:34: error: no matching member function for call to 'ToNumber'
                        Coroutine::pool_size = value->ToNumber()->Value();
                                               ~~~~~~~^~~~~~~~
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:2366:44: note: candidate function not viable: requires single argument 'context', but no arguments were provided
  V8_WARN_UNUSED_RESULT MaybeLocal<Number> ToNumber(
                                           ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:2383:35: note: candidate function not viable: requires single argument 'isolate', but no arguments were provided
                    Local<Number> ToNumber(Isolate* isolate) const);
                                  ^
../src/fibers.cc:793:72: error: too few arguments to function call, expected at least 3, have 2
                        fn->SetAccessor(uni::NewLatin1Symbol(isolate, \"current\"), GetCurrent);
                        ~~~~~~~~~~~~~~~                                                     ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:3214:3: note: 'SetAccessor' declared here
  V8_WARN_UNUSED_RESULT Maybe<bool> SetAccessor(
  ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8config.h:410:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/fibers.cc:794:63: error: no viable conversion from 'uni::FunctionType (Local<v8::String>, const uni::GetterCallbackInfo &)' (aka 'void (Local<v8::String>, const PropertyCallbackInfo<v8::Value> &)') to 'Local<v8::Name>'
                        fn->SetAccessor(uni::NewLatin1Symbol(isolate, \"poolSize\"), GetPoolSize, SetPoolSize);
                                                                                   ^~~~~~~~~~~
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:202:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'uni::FunctionType (Local<v8::String>, const uni::GetterCallbackInfo &)' (aka 'void (Local<v8::String>, const PropertyCallbackInfo<v8::Value> &)') to 'const v8::Local<v8::Name> &' for 1st argument
class Local {
      ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:202:7: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'uni::FunctionType (Local<v8::String>, const uni::GetterCallbackInfo &)' (aka 'void (Local<v8::String>, const PropertyCallbackInfo<v8::Value> &)') to 'v8::Local<v8::Name> &&' for 1st argument
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:206:13: note: candidate template ignored: could not match 'Local<type-parameter-0-0>' against 'void (*)(v8::Local<v8::String>, const v8::PropertyCallbackInfo<v8::Value> &)'
  V8_INLINE Local(Local<S> that)
            ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:3215:43: note: passing argument to parameter 'name' here
      Local<Context> context, Local<Name> name,
                                          ^
../src/fibers.cc:795:84: error: too few arguments to function call, expected at least 3, have 2
                        fn->SetAccessor(uni::NewLatin1Symbol(isolate, \"fibersCreated\"), GetFibersCreated);
                        ~~~~~~~~~~~~~~~                                                                 ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:3214:3: note: 'SetAccessor' declared here
  V8_WARN_UNUSED_RESULT Maybe<bool> SetAccessor(
  ^
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8config.h:410:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
In file included from ../src/fibers.cc:1:
In file included from ../src/coroutine.h:1:
In file included from /Users/akos.kitta/.node-gyp/10.3.0/include/node/node.h:63:
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:213:5: error: assigning to 'v8::Context *' from incompatible type 'v8::String *'
    TYPE_CHECK(T, S);
    ^~~~~~~~~~~~~~~~
/Users/akos.kitta/.node-gyp/10.3.0/include/node/v8.h:167:39: note: expanded from macro 'TYPE_CHECK'
    *(static_cast<T* volatile*>(0)) = static_cast<S*>(0);      \\
                                      ^~~~~~~~~~~~~~~~~~
../src/fibers.cc:794:20: note: in instantiation of function template specialization 'v8::Local<v8::Context>::Local<v8::String>' requested here
                        fn->SetAccessor(uni::NewLatin1Symbol(isolate, \"poolSize\"), GetPoolSize, SetPoolSize);
                                        ^
2 warnings and 9 errors generated.
make: *** [Release/obj.target/fibers/src/fibers.o] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/Users/akos.kitta/Desktop/theia/node_modules/node-gyp/lib/build.js:258:23)
gyp ERR! stack     at ChildProcess.emit (events.js:182:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:237:12)
gyp ERR! System Darwin 17.4.0
gyp ERR! command \"/Users/akos.kitta/.nvm/versions/node/v10.3.0/bin/node\" \"/Users/akos.kitta/Desktop/theia/node_modules/.bin/node-gyp\" \"rebuild\" \"--release\"
gyp ERR! cwd /Users/akos.kitta/Desktop/theia/node_modules/fibers
gyp ERR! node -v v10.3.0
gyp ERR! node-gyp -v v3.6.2
gyp ERR! not ok 
node-gyp exited with code: 1
Please make sure you are using a supported platform and node version. If you
would like to compile fibers on this machine please make sure you have setup your
build environment--
Windows + OS X instructions here: https://github.com/nodejs/node-gyp
Ubuntu users please run: `sudo apt-get install g++ build-essential`
```

### PR DESCRIPTION
**Pull request by jankeromnes**
Fixes #2009.

This is #3768, but:
- with the changes suggested in https://github.com/theia-ide/theia/pull/3768#issuecomment-446207160
- without the "DROPME" commit
- with `yarn.lock` where only the 6 relevant dependencies were upgraded

Since I couldn't contribute to #3768, I've created a new pull request, but preserved the `Co-authored-by:` information.

### CODE REVIEW

**Comment by kittaakos** [comment]
@jankeromnes, I think we should see a successful build on both Travis and AppVeyor with Node.js 10. Then we can revert that commit.

**Comment by jankeromnes** [comment]
I've confirmed that this works in Gitpod by running:

```bash
nvm install 10
npm config set python /usr/bin/python --global
npm config set python /usr/bin/python
npm install -g typescript yarn
```

(in an attempt to emulate this upgrade to node 10: https://github.com/gitpod-io/workspace-images/pull/70)

and then `yarn` to verify that it builds

and then started Theia with `cd examples/browser && yarn run start ../.. --hostname 0.0.0.0` to confirm that Theia starts up and can be used normally, without suspicious errors in the browser console or in the server logs.

**Comment by jankeromnes** [comment]
> I think we should see a successful build on both Travis and AppVeyor with Node.js 10. Then we can revert that commit.

Do I need to do anything in particular to make Travis and AppVeryor try with Node.js 10?

I'll also need to test the Electron app locally.

**Comment by kittaakos** [comment]
> Do I need to do anything in particular to make Travis and AppVeryor try with Node.js 10?

https://github.com/theia-ide/theia/blob/2bd0e8f4e9f8ba84fe9a10085232c444b8b215a8/appveyor.yml#L5

https://github.com/theia-ide/theia/blob/2bd0e8f4e9f8ba84fe9a10085232c444b8b215a8/.travis.yml#L3

> I'll also need to test the Electron app locally.

No. Electron comes with its own Node.js. It is sufficient if the CIs pass.

**Comment by kittaakos** [comment]
And the builds ~are~ were failing due to this: https://github.com/theia-ide/theia/issues/3787 Some details are [here](https://github.com/Microsoft/vscode/issues/28434#issuecomment-346199674).

**Comment by jankeromnes** [comment]
AppVeyor is green! 💚

However Travis failed with:

```
lerna ERR!   1 failing
lerna ERR! 
lerna ERR!   1) ripgrep-search-in-workspace-server fails gracefully when rg isn't executable:
lerna ERR!      TypeError: Cannot read property 'on' of undefined
lerna ERR!       at RipgrepSearchInWorkspaceServer.search (lib/node/ripgrep-search-in-workspace-server.js:31:79)
lerna ERR!       at lib/node/ripgrep-search-in-workspace-server.slow-spec.js:53:619
lerna ERR!       at new Promise (<anonymous>)
lerna ERR!       at Context.<anonymous> (lib/node/ripgrep-search-in-workspace-server.slow-spec.js:52:4850)
lerna ERR!       at step (lib/node/ripgrep-search-in-workspace-server.slow-spec.js:15:56451)
lerna ERR!       at Object.next (lib/node/ripgrep-search-in-workspace-server.slow-spec.js:15:53580)
lerna ERR!       at lib/node/ripgrep-search-in-workspace-server.slow-spec.js:15:52746
lerna ERR!       at new Promise (<anonymous>)
lerna ERR!       at __awaiter (lib/node/ripgrep-search-in-workspace-server.slow-spec.js:15:51901)
lerna ERR!       at Context.<anonymous> (lib/node/ripgrep-search-in-workspace-server.slow-spec.js:52:4577)
```

Maybe some API change? The only occurrences of `.on` I can find in `ripgrep-search-in-workspace-server.ts` are:

https://github.com/theia-ide/theia/blob/2bd0e8f4e9f8ba84fe9a10085232c444b8b215a8/packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts#L160

and

https://github.com/theia-ide/theia/blob/2bd0e8f4e9f8ba84fe9a10085232c444b8b215a8/packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts#L227

Maybe `process.output` is `undefined` for some reason?

**Comment by jankeromnes** [comment]
`process.output` can be set to `undefined` either by:

https://github.com/theia-ide/theia/blob/2bd0e8f4e9f8ba84fe9a10085232c444b8b215a8/packages/process/src/node/raw-process.ts#L122

or by:

https://github.com/theia-ide/theia/blob/2bd0e8f4e9f8ba84fe9a10085232c444b8b215a8/packages/process/src/node/raw-process.ts#L134

(or maybe it isn't set at all?)

File: .travis.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 sudo: required
 language: node_js
-node_js: '8'
+node_js: '10'
```
Comment (akosyakov) [inline .travis.yml]: So our base line now nodejs 10? If not it should stay nodejs 8, we should avoid using any nodejs 10 apis

Comment (jankeromnes) [reply .travis.yml]: It's just a test to verify that the other commit fixes Theia for Node 10. I'll revert it once the tests pass.

**Comment by jankeromnes** [comment]
Rebased in order to retrigger the tests, and see if the error reproduces.

Also, it looks very suspicious to me that we do this when `catch (error)`:

https://github.com/theia-ide/theia/blob/2bd0e8f4e9f8ba84fe9a10085232c444b8b215a8/packages/process/src/node/raw-process.ts#L131-L142

But only this when `process.on('error')`:

https://github.com/theia-ide/theia/blob/2bd0e8f4e9f8ba84fe9a10085232c444b8b215a8/packages/process/src/node/raw-process.ts#L108-L112

(Wouldn't the streams stay undefined in that second case?)

**Comment by marechal-p** [comment]
@jankeromnes in the second case streams wouldn't be undefined, this is part of the way asynchronous APIs work in NodeJS: when you register callbacks, they will only be called on the next tick, unlike errors thrown and caught synchronously (first case).

**Comment by akosyakov** [comment]
@jankeromnes Would it be fined to close in favour of https://github.com/theia-ide/theia/pull/4750?

**Comment by jankeromnes** [comment]
> Would it be fined to close in favour of #4750?

Yes please! Thank you @thegecko for addressing this. 👍

**Pull request closed without merging**